### ISSUE TITLE
Doc2Dataset fails if lineage sources are specified as UUIDs rather than strings

### ISSUE THREAD

**Issue by tijmenr** [opendatacube/datacube-core#1300]
Converting an input doc that contains lineage references to a Dataset (using `Doc2Dataset(index, fail_on_missing_lineage=True, verify_lineage=False, skip_lineage=False)`) fails if the lineage references in the input doc are of type `UUID` instead of `str`.

In [dataset_resolver, hl.py around line 174](https://github.com/opendatacube/datacube-core/blob/c5486c854dc4fe3e1bef670a5ce01010d0caa938/datacube/index/hl.py#L174), the identifiers of the resolved lineage sources are converted to strings. Then, missing lineage items are determined by subtracting the resolved identifiers from the input set of lineage sources. But if the input references are of type `UUID`, this simple set subtraction is incorrect, and as a result the lineage items will be considered missing: 
```python
        db_dss = {str(ds.id): ds for ds in index.datasets.bulk_get(all_uuid)}

        lineage_uuids = set(filter(lambda x: x != main_uuid, all_uuid))
        missing_lineage = lineage_uuids - set(db_dss)
```
(After that, also generating the error message in line 180 fails, because it expects missing_lineage items to be strings).

In practice, it is very likely that lineage source identifiers will be of type `UUID`, given that the `Dataset` class uses `UUID`s for identifiers, as do external interfaces like e.g. `eodatasets3.DatasetPrepare.note_source_datasets()`.
Before doing the subtraction, both lineage_uuids and db_dss sets should be normalized to contain a single datatype.

**Comment by Kirill888** [comment]
This was mostly(only?) used with documents that were just loaded from yaml/json, so not surprising that type normalization of ids was missed. What is the source of documents in your case? 

@tijmenr seems like an easy fix, want to raise a PR?

**Comment by Kirill888** [comment]
But I also think that `eodatasets3` should not put raw UUID objects into those docs as they are not json serializable by default

```python
import uuid
import json
json.dumps([uuid.uuid4()]) #<- raises TypeError
```

so adding those to db will probably fail anyway...

**Comment by tijmenr** [comment]
In our use case, documents are generated using the `DatasetAssembler` class from `eodatasets3`, and its `note_source_datasets` function actually forces the source ids into UUIDs.
Eodatasets3 mainly uses YAML format, and they convert UUIDs when serializing (see https://github.com/opendatacube/eo-datasets/blob/develop/eodatasets3/serialise.py).

I created a PR that probably fixes the issue, but note that I haven't had time to extensively test this (I only did from the perspective of our use case).
